File: internal/store/projects_test.go
package store

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/jwstover/tend/internal/store/gen"
	"github.com/jwstover/tend/internal/task"
)

func TestProjectCRUD(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "  tend  ")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	if p.Name != "tend" {
		t.Errorf("Name = %q, want the trimmed %q", p.Name, "tend")
	}

	if _, err := s.CreateProject(ctx, "   "); !errors.Is(err, task.ErrEmptyProjectName) {
		t.Errorf("CreateProject(blank) = %v, want ErrEmptyProjectName", err)
	}
	// Unique is NOCASE, so this is the same project.
	if _, err := s.CreateProject(ctx, "TEND"); err == nil {
		t.Error("CreateProject with a case-variant duplicate should fail")
	}

	// ...and lookup is case-insensitive in the same way.
	found, err := s.ProjectByName(ctx, "TeNd")
	if err != nil {
		t.Fatalf("ProjectByName: %v", err)
	}
	if found.ID != p.ID {
		t.Errorf("ProjectByName resolved to %d, want %d", found.ID, p.ID)
	}
	if _, err := s.ProjectByName(ctx, "nope"); !errors.Is(err, task.ErrProjectNotFound) {
		t.Errorf("ProjectByName(unknown) = %v, want ErrProjectNotFound", err)
	}

	if err := s.RenameProject(ctx, p.ID, "tend-cli"); err != nil {
		t.Fatalf("RenameProject: %v", err)
	}
	got, err := s.GetProject(ctx, p.ID)
	if err != nil {
		t.Fatalf("GetProject: %v", err)
	}
	if got.Name != "tend-cli" {
		t.Errorf("Name after rename = %q", got.Name)
	}

	if err := s.SetProjectArchived(ctx, p.ID, true); err != nil {
		t.Fatalf("SetProjectArchived: %v", err)
	}
	if got, _ = s.GetProject(ctx, p.ID); !got.Archived() {
		t.Error("project should read archived")
	}
	if err := s.SetProjectArchived(ctx, p.ID, false); err != nil {
		t.Fatalf("SetProjectArchived(false): %v", err)
	}
	if got, _ = s.GetProject(ctx, p.ID); got.Archived() {
		t.Error("project should read unarchived")
	}
}

// project_id carries no foreign key (migration 00007), so the store is
// what stops a delete from orphaning tasks. This is that guarantee.
func TestDeleteProjectReassignsItsTasks(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "doomed")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	stranded, err := s.AddTaskIn(ctx, p.ID, "do not orphan me")
	if err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}

	if err := s.DeleteProject(ctx, p.ID); err != nil {
		t.Fatalf("DeleteProject: %v", err)
	}
	got, err := s.GetTask(ctx, stranded.ID)
	if err != nil {
		t.Fatalf("GetTask: %v", err)
	}
	if got.ProjectID != task.DefaultProjectID {
		t.Errorf("ProjectID after its project was deleted = %d, want the default %d",
			got.ProjectID, task.DefaultProjectID)
	}
	if _, err := s.GetProject(ctx, p.ID); !errors.Is(err, task.ErrProjectNotFound) {
		t.Errorf("GetProject(deleted) = %v, want ErrProjectNotFound", err)
	}
}

func TestDeleteDefaultProjectRefused(t *testing.T) {
	s := newTestStore(t)
	if err := s.DeleteProject(context.Background(), task.DefaultProjectID); !errors.Is(err, task.ErrProtectedProject) {
		t.Errorf("DeleteProject(default) = %v, want ErrProtectedProject", err)
	}
}

func TestListProjectsCountsLiveTopLevelTasks(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "counted")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	parent, err := s.AddTaskIn(ctx, p.ID, "top level")
	if err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}
	// A sub-task must not inflate the count: the number beside a project
	// is what selecting it puts on screen as rows.
	if _, err := s.AddChild(ctx, parent.ID, "sub"); err != nil {
		t.Fatalf("AddChild: %v", err)
	}
	// Nor should a completed one, which the live view filters out.
	done, err := s.AddTaskIn(ctx, p.ID, "finished")
	if err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}
	if err := s.SetState(ctx, done.ID, task.StateDone); err != nil {
		t.Fatalf("SetState: %v", err)
	}

	projects, err := s.ListProjects(ctx)
	if err != nil {
		t.Fatalf("ListProjects: %v", err)
	}
	for _, got := range projects {
		if got.ID != p.ID {
			continue
		}
		if got.LiveCount != 1 {
			t.Errorf("LiveCount = %d, want 1 (top-level and live only)", got.LiveCount)
		}
		return
	}
	t.Fatalf("project %d missing from ListProjects", p.ID)
}

// The setting is the TUI's memory of where it was, not a capture target.
func TestActiveProjectRoundTripAndFallbacks(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	// Never set: the default, not an error.
	got, err := s.ActiveProjectID(ctx)
	if err != nil || got != task.DefaultProjectID {
		t.Fatalf("ActiveProjectID on a fresh db = (%d, %v), want the default", got, err)
	}

	p, err := s.CreateProject(ctx, "active")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	if err := s.SetActiveProject(ctx, p.ID); err != nil {
		t.Fatalf("SetActiveProject: %v", err)
	}
	if got, _ = s.ActiveProjectID(ctx); got != p.ID {
		t.Errorf("ActiveProjectID = %d, want %d", got, p.ID)
	}

	// Deleting the remembered project must not strand the TUI's restore.
	if err := s.DeleteProject(ctx, p.ID); err != nil {
		t.Fatalf("DeleteProject: %v", err)
	}
	if got, err = s.ActiveProjectID(ctx); err != nil || got != task.DefaultProjectID {
		t.Errorf("ActiveProjectID after its project was deleted = (%d, %v), want the default", got, err)
	}

	// A corrupt stored value degrades the same way rather than failing: a
	// bad UI preference must not stop the TUI from starting.
	if err := s.q.SetSetting(ctx, gen.SetSettingParams{
		Key: settingActiveProject, Value: "not a number",
	}); err != nil {
		t.Fatalf("seeding a corrupt setting: %v", err)
	}
	if got, err = s.ActiveProjectID(ctx); err != nil || got != task.DefaultProjectID {
		t.Errorf("ActiveProjectID with a corrupt value = (%d, %v), want the default", got, err)
	}
}

func TestSetProjectMovesWholeSubtree(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	dest, err := s.CreateProject(ctx, "destination")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	root, err := s.AddTask(ctx, "root")
	if err != nil {
		t.Fatalf("AddTask: %v", err)
	}
	child, err := s.AddChild(ctx, root.ID, "child")
	if err != nil {
		t.Fatalf("AddChild: %v", err)
	}
	// Three levels deep: the walk replaces a recursive CTE, so depth is
	// exactly what could regress.
	grandchild, err := s.AddChild(ctx, child.ID, "grandchild")
	if err != nil {
		t.Fatalf("AddChild: %v", err)
	}

	if err := s.SetProject(ctx, root.ID, dest.ID); err != nil {
		t.Fatalf("SetProject: %v", err)
	}
	for _, id := range []int64{root.ID, child.ID, grandchild.ID} {
		got, err := s.GetTask(ctx, id)
		if err != nil {
			t.Fatalf("GetTask(%d): %v", id, err)
		}
		if got.ProjectID != dest.ID {
			t.Errorf("task %d ProjectID = %d, want %d (the whole sub-tree moves)",
				id, got.ProjectID, dest.ID)
		}
	}
}

// A sub-task can never sit in a different project from its parent, so the
// insert copies the parent's project rather than defaulting.
func TestAddChildInheritsParentProject(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "parent project")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	parent, err := s.AddTaskIn(ctx, p.ID, "parent")
	if err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}
	child, err := s.AddChild(ctx, parent.ID, "child")
	if err != nil {
		t.Fatalf("AddChild: %v", err)
	}
	if child.ProjectID != p.ID {
		t.Errorf("child ProjectID = %d, want the parent's %d", child.ProjectID, p.ID)
	}
}

func TestListLiveScopesToProject(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "scoped")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	if _, err := s.AddTaskIn(ctx, p.ID, "in scope"); err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}
	if _, err := s.AddTaskIn(ctx, task.DefaultProjectID, "out of scope"); err != nil {
		t.Fatalf("AddTaskIn: %v", err)
	}

	all, err := s.ListLive(ctx, nil)
	if err != nil {
		t.Fatalf("ListLive(nil): %v", err)
	}
	if len(all) != 2 {
		t.Errorf("ListLive(nil) returned %d tasks, want both", len(all))
	}

	scoped, err := s.ListLive(ctx, &p.ID)
	if err != nil {
		t.Fatalf("ListLive(scoped): %v", err)
	}
	if len(scoped) != 1 || scoped[0].Title != "in scope" {
		t.Errorf("ListLive(scoped) = %+v, want only the in-scope task", scoped)
	}

	n, err := s.CountInbox(ctx, &p.ID)
	if err != nil {
		t.Fatalf("CountInbox: %v", err)
	}
	if n != 1 {
		t.Errorf("CountInbox(scoped) = %d, want 1", n)
	}
}

// A bare capture never consults the remembered project. The shell has no
// hidden state steering it, so `tend add` behaves identically in every
// terminal regardless of what the TUI is looking at.
func TestAddTaskIgnoresTheRememberedProject(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	p, err := s.CreateProject(ctx, "elsewhere")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	if err := s.SetActiveProject(ctx, p.ID); err != nil {
		t.Fatalf("SetActiveProject: %v", err)
	}

	captured, err := s.AddTask(ctx, "bare capture")
	if err != nil {
		t.Fatalf("AddTask: %v", err)
	}
	if captured.ProjectID != task.DefaultProjectID {
		t.Errorf("AddTask landed in %d, want the default project %d",
			captured.ProjectID, task.DefaultProjectID)
	}

	withBody, err := s.AddTaskWithBody(ctx, "with a body", "context")
	if err != nil {
		t.Fatalf("AddTaskWithBody: %v", err)
	}
	if withBody.ProjectID != task.DefaultProjectID {
		t.Errorf("AddTaskWithBody landed in %d, want the default project %d",
			withBody.ProjectID, task.DefaultProjectID)
	}
}

// Moving a task logs exactly one event, not one per row in its sub-tree:
// the whole point of writing it in the store rather than in a trigger.
func TestSetProjectLogsOneEventForTheWholeSubtree(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	dest, err := s.CreateProject(ctx, "destination")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	root, err := s.AddTask(ctx, "root")
	if err != nil {
		t.Fatalf("AddTask: %v", err)
	}
	child, err := s.AddChild(ctx, root.ID, "child")
	if err != nil {
		t.Fatalf("AddChild: %v", err)
	}
	if _, err := s.AddChild(ctx, child.ID, "grandchild"); err != nil {
		t.Fatalf("AddChild: %v", err)
	}

	before := projectEvents(t, s)
	if err := s.SetProject(ctx, root.ID, dest.ID); err != nil {
		t.Fatalf("SetProject: %v", err)
	}
	got := projectEvents(t, s)
	if len(got) != len(before)+1 {
		t.Fatalf("logged %d project events for one move, want 1", len(got)-len(before))
	}

	ev := got[len(got)-1]
	if ev.TaskID != root.ID {
		t.Errorf("event is for task %d, want the moved root %d", ev.TaskID, root.ID)
	}
	if ev.Old == nil || *ev.Old != "Unsorted" || ev.New == nil || *ev.New != "destination" {
		t.Errorf("event = (%v -> %v), want Unsorted -> destination", ev.Old, ev.New)
	}
}

// Moving a task to where it already is changes nothing and logs nothing,
// mirroring the state trigger's OLD <> NEW guard.
func TestSetProjectToTheSameProjectIsANoop(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	created, err := s.AddTask(ctx, "staying put")
	if err != nil {
		t.Fatalf("AddTask: %v", err)
	}
	before := projectEvents(t, s)
	if err := s.SetProject(ctx, created.ID, task.DefaultProjectID); err != nil {
		t.Fatalf("SetProject: %v", err)
	}
	if got := projectEvents(t, s); len(got) != len(before) {
		t.Errorf("a no-op move logged %d event(s)", len(got)-len(before))
	}
}

// The event snapshots project names, so the log stays readable after the
// project it names is renamed or deleted -- the same reason task_events
// snapshots task_title.
func TestProjectEventSurvivesRenamingTheProject(t *testing.T) {
	s := newTestStore(t)
	ctx := context.Background()

	dest, err := s.CreateProject(ctx, "original name")
	if err != nil {
		t.Fatalf("CreateProject: %v", err)
	}
	created, err := s.AddTask(ctx, "moved")
	if err != nil {
		t.Fatalf("AddTask: %v", err)
	}
	if err := s.SetProject(ctx, created.ID, dest.ID); err != nil {
		t.Fatalf("SetProject: %v", err)
	}
	if err := s.RenameProject(ctx, dest.ID, "renamed later"); err != nil {
		t.Fatalf("RenameProject: %v", err)
	}

	got := projectEvents(t, s)
	last := got[len(got)-1]
	if last.New == nil || *last.New != "original name" {
		t.Errorf("event New = %v, want the name snapshotted at move time", last.New)
	}
}

// projectEvents returns every project-kind event in a wide window.
func projectEvents(t *testing.T, s *Store) []task.Event {
	t.Helper()
	events, err := s.ListEvents(context.Background(),
		time.Now().Add(-24*time.Hour), time.Now().Add(24*time.Hour))
	if err != nil {
		t.Fatalf("ListEvents: %v", err)
	}
	var out []task.Event
	for _, ev := range events {
		if ev.Kind == task.EventProject {
			out = append(out, ev)
		}
	}
	return out
}

File: elixir/test/tend/store/projects_test.exs
defmodule Tend.Store.ProjectsTest do
  @moduledoc """
  The project rows and the `active_project_id` setting, against a real temp
  database.

  A port of the project-CRUD, delete-reassignment, live-count and
  active-project cases in `internal/store/projects_test.go`. The four
  `SetProject`/task-move cases there are `Tend.Store.HierarchyTest`'s; the
  three cases exercising `list_live/2`, `add_child/3` and `add_task/2` against
  a project other than the default are also `Tend.Store.HierarchyTest`'s,
  since they are really about the hierarchy and capture surfaces, not this
  one.
  """

  use ExUnit.Case, async: true

  alias Tend.Error
  alias Tend.Store
  alias Tend.Task.Project
  alias Tend.Test.SQL

  @moduletag :tmp_dir

  setup %{tmp_dir: dir} do
    {:ok, store} = Store.open(Path.join(dir, "tend.db"))
    on_exit(fn -> Store.close(store) end)
    %{store: store}
  end

  defp project!(store, name) do
    {:ok, project} = Store.create_project(store, name)
    project
  end

  test "create_project/2 trims the name, get_project/2 and project_by_name/2 read it back",
       %{store: store} do
    {:ok, project} = Store.create_project(store, "  tend  ")
    assert project.name == "tend"

    assert Store.create_project(store, "   ") == {:error, :empty_project_name}

    # Unique is NOCASE, so this is the same project: the driver's constraint
    # error, not a sentinel.
    assert {:error, {:query_failed, "creating project \"TEND\"", _}} =
             Store.create_project(store, "TEND")

    assert {:ok, found} = Store.project_by_name(store, "TeNd")
    assert found.id == project.id

    assert Store.project_by_name(store, "nope") == {:error, {:project_not_found, "nope"}}

    assert Error.message({:project_not_found, "nope"}) == ~s|project "nope": project not found|
  end

  test "get_project/2 on an id that is not there", %{store: store} do
    assert Store.get_project(store, 999) == {:error, {:project_not_found, 999}}
    assert Error.message({:project_not_found, 999}) == "project 999: project not found"
  end

  test "rename_project/3 and set_project_archived/3 round-trip through get_project/2", %{
    store: store
  } do
    project = project!(store, "tend")

    assert Store.rename_project(store, project.id, "tend-cli") == :ok
    assert {:ok, got} = Store.get_project(store, project.id)
    assert got.name == "tend-cli"
    refute Project.archived?(got)

    assert Store.set_project_archived(store, project.id, true) == :ok
    assert {:ok, got} = Store.get_project(store, project.id)
    assert Project.archived?(got)
    assert got.archived_at != nil

    assert Store.set_project_archived(store, project.id, false) == :ok
    assert {:ok, got} = Store.get_project(store, project.id)
    refute Project.archived?(got)
    assert got.archived_at == nil
  end

  # project_id carries no foreign key (migration 00007), so the store is what
  # stops a delete from orphaning tasks. This is that guarantee.
  test "delete_project/2 reassigns its tasks to the default project", %{store: store} do
    project = project!(store, "doomed")
    {:ok, stranded} = Store.add_task_in(store, project.id, "do not orphan me")

    assert Store.delete_project(store, project.id) == :ok

    assert {:ok, got} = Store.get_task(store, stranded.id)
    assert got.project_id == Project.default_id()

    assert Store.get_project(store, project.id) == {:error, {:project_not_found, project.id}}
  end

  test "delete_project/2 refuses the default project", %{store: store} do
    assert Store.delete_project(store, Project.default_id()) == {:error, :protected_project}
  end

  test "list_projects/1 counts only live top-level tasks", %{store: store} do
    project = project!(store, "counted")
    {:ok, parent} = Store.add_task_in(store, project.id, "top level")

    # A sub-task must not inflate the count: the number beside a project is
    # what selecting it puts on screen as rows.
    {:ok, _child} = Store.add_child(store, parent.id, "sub")

    # Nor should a completed one, which the live view filters out.
    {:ok, done} = Store.add_task_in(store, project.id, "finished")
    :ok = Store.set_state(store, done.id, :done)

    assert {:ok, projects} = Store.list_projects(store)
    assert %Project{live_count: 1} = Enum.find(projects, &(&1.id == project.id))
  end

  test "list_projects/1 includes archived projects, ordered by sort_order then name", %{
    store: store
  } do
    archived = project!(store, "archived one")
    :ok = Store.set_project_archived(store, archived.id, true)

    zeta = project!(store, "zeta")
    alpha = project!(store, "alpha")

    assert {:ok, projects} = Store.list_projects(store)
    assert Enum.any?(projects, &(&1.id == archived.id and Project.archived?(&1)))

    # Every project here shares sort_order 0 (nothing in this port yet sets
    # it), so the tie-break is name -- alpha before zeta regardless of
    # creation order.
    ids = Enum.map(projects, & &1.id)
    assert Enum.find_index(ids, &(&1 == alpha.id)) < Enum.find_index(ids, &(&1 == zeta.id))
  end

  test "active_project_id/1 and set_active_project/2 round-trip, with fallbacks", %{
    store: store
  } do
    # Never set: the default, not an error.
    assert Store.active_project_id(store) == {:ok, Project.default_id()}

    project = project!(store, "active")
    assert Store.set_active_project(store, project.id) == :ok
    assert Store.active_project_id(store) == {:ok, project.id}

    # Deleting the remembered project must not strand the TUI's restore.
    assert Store.delete_project(store, project.id) == :ok
    assert Store.active_project_id(store) == {:ok, Project.default_id()}

    # A corrupt stored value degrades the same way rather than failing: a bad
    # UI preference must not stop the TUI from starting. There is no public
    # setter for a non-numeric value, so this seeds the row directly -- the
    # row already exists from set_active_project/2 above, so it is an UPDATE.
    SQL.exec!(
      store.conn,
      "UPDATE settings SET value = 'not a number' WHERE key = 'active_project_id'"
    )

    assert Store.active_project_id(store) == {:ok, Project.default_id()}
  end
end
